File: backend/app/engine/sizing.py
"""Dynamic position sizing.

    lots = floor( (capital × risk% ) / (stop_loss_points × lot_size) )

The raw risk-derived lot count is then clamped by two hard realities: you cannot
buy more premium than the account holds, and the exchange only trades whole
lots.  NSE index lot increments (NIFTY 25/50/75, BANKNIFTY 15, FINNIFTY 25/40)
are respected via the scrip master's lot size, with the configured increments as
the fallback ladder.
"""

from __future__ import annotations

import math

from ..config import INDEX_UNIVERSE, settings
from ..schemas import SizingResult


def round_to_lot(quantity: int, lot_size: int) -> int:
    """Floor an arbitrary quantity onto the nearest tradable lot multiple."""
    if lot_size <= 0:
        return max(0, quantity)
    return (quantity // lot_size) * lot_size


def resolve_lot_size(underlying: str, override: int | None = None) -> int:
    if override and override > 0:
        return int(override)
    return int(INDEX_UNIVERSE.get(underlying, {}).get("lot_size", 1))


def calculate_size(
    underlying: str,
    stop_loss_points: float,
    capital: float | None = None,
    risk_pct: float | None = None,
    lot_size: int | None = None,
    entry_price: float | None = None,
    max_deployable: float | None = None,
) -> SizingResult:
    """Compute the risk-adjusted lot count for one Delta-K entry.

    Args:
        stop_loss_points: Premium points between entry and stop.
        capital: Account capital; defaults to the paper capital setting.
        risk_pct: Percent of capital risked on this trade.
        lot_size: Contract multiplier; defaults to the index lot size.
        entry_price: Option premium, used for the entry-cost/affordability cap.
        max_deployable: Optional ceiling on deployable capital (free margin).
    """
    capital = float(capital if capital is not None else settings.paper_capital)
    risk_pct = float(risk_pct if risk_pct is not None else settings.risk_pct_per_trade)
    lot = resolve_lot_size(underlying, lot_size)

    risk_amount = capital * (risk_pct / 100.0)
    risk_per_lot = max(0.0, stop_loss_points) * lot

    capped_by: str | None = None
    if risk_per_lot <= 0:
        lots = 0
        capped_by = "INVALID_STOP"
    else:
        lots = int(math.floor(risk_amount / risk_per_lot))

    entry_cost = 0.0
    if entry_price and entry_price > 0 and lots > 0:
        budget = capital if max_deployable is None else min(capital, max_deployable)
        cost_per_lot = entry_price * lot
        affordable = int(math.floor(budget / cost_per_lot)) if cost_per_lot > 0 else 0
        if affordable < lots:
            lots = affordable
            capped_by = "CAPITAL"
        entry_cost = lots * cost_per_lot

    lots = max(0, lots)
    if lots == 0 and capped_by is None:
        capped_by = "RISK_BUDGET"

    return SizingResult(
        lots=lots,
        quantity=lots * lot,
        lot_size=lot,
        risk_amount=round(risk_amount, 2),
        risk_per_lot=round(risk_per_lot, 2),
        entry_cost=round(entry_cost, 2),
        capital=round(capital, 2),
        risk_pct=risk_pct,
        capped_by=capped_by,
    )

File: web/lib/engine/sizing.ts
import type { Side, SizingResult } from "@/lib/types";
import { INDEX_UNIVERSE } from "./config";

/**
 * Dynamic position sizing — port of `backend/app/engine/sizing.py`.
 *
 *     lots = floor( (capital × risk%) / (stop_loss_points × lot_size) )
 *
 * The risk-derived count is then clamped by two hard realities: you cannot buy
 * more premium than the account holds, and the exchange only trades whole lots.
 */

export function roundToLot(quantity: number, lotSize: number): number {
  if (lotSize <= 0) return Math.max(0, quantity);
  return Math.floor(quantity / lotSize) * lotSize;
}

export function resolveLotSize(underlying: string, override?: number): number {
  if (override && override > 0) return Math.trunc(override);
  return INDEX_UNIVERSE[underlying]?.lotSize ?? 1;
}

export function calculateSize(params: {
  underlying: string;
  stopLossPoints: number;
  capital: number;
  riskPct: number;
  lotSize?: number;
  entryPrice?: number;
  maxDeployable?: number;
  /**
   * Hard ceiling on this position's premium spend, percent of `capital`,
   * independent of the risk-% math above. Without it, a generous risk% on a
   * cheaply-stopped contract sizes up until the capital-affordability clamp
   * binds — which can mean "spend nearly all deployable capital" while the
   * panel still reads "risk N%". This bounds that directly.
   */
  maxPositionCapitalPct?: number;
}): SizingResult {
  const {
    underlying,
    stopLossPoints,
    capital,
    riskPct,
    entryPrice,
    maxDeployable,
    maxPositionCapitalPct,
  } = params;

  const lot = resolveLotSize(underlying, params.lotSize);
  const riskAmount = capital * (riskPct / 100);
  const riskPerLot = Math.max(0, stopLossPoints) * lot;

  let cappedBy: string | null = null;
  let lots = 0;

  if (riskPerLot <= 0) {
    cappedBy = "INVALID_STOP";
  } else {
    lots = Math.floor(riskAmount / riskPerLot);
  }
  // What the risk-% math alone would size this to, before the
  // affordability clamp below can shrink it — the baseline
  // `singleLotConstrained` compares the final lot count against.
  const riskLots = lots;

  let entryCost = 0;
  if (entryPrice && entryPrice > 0 && lots > 0) {
    const deployableBudget =
      maxDeployable === undefined ? capital : Math.min(capital, maxDeployable);
    const concentrationBudget =
      maxPositionCapitalPct && maxPositionCapitalPct > 0
        ? capital * (maxPositionCapitalPct / 100)
        : Infinity;
    const budget = Math.min(deployableBudget, concentrationBudget);
    const costPerLot = entryPrice * lot;
    const affordable = costPerLot > 0 ? Math.floor(budget / costPerLot) : 0;
    if (affordable < lots) {
      lots = affordable;
      cappedBy = concentrationBudget < deployableBudget ? "CONCENTRATION" : "CAPITAL";
    }
    entryCost = lots * costPerLot;
  }

  lots = Math.max(0, lots);
  if (lots === 0 && cappedBy === null) cappedBy = "RISK_BUDGET";

  // The risk math wanted more than one lot, but capital/concentration
  // clamped it down to exactly one — not "the risk math itself only ever
  // wanted 1 lot," which needs no special handling downstream.
  const singleLotConstrained =
    lots === 1 && riskLots > 1 && (cappedBy === "CAPITAL" || cappedBy === "CONCENTRATION");

  const r2 = (n: number) => Number(n.toFixed(2));
  return {
    lots,
    quantity: lots * lot,
    lot_size: lot,
    risk_amount: r2(riskAmount),
    risk_per_lot: r2(riskPerLot),
    entry_cost: r2(entryCost),
    capital: r2(capital),
    risk_pct: riskPct,
    single_lot_constrained: singleLotConstrained,
    capped_by: cappedBy,
  };
}

/* ----------------------------------------------------- portfolio risk cap */

/**
 * Enough of a leg to price its own worst case — a live `Position` and a
 * not-yet-opened candidate both satisfy this.
 */
export interface RiskLeg {
  side: Side;
  avg_price: number;
  stop_loss: number | null;
  quantity: number;
}

/**
 * Rupees lost if this leg's own stop is hit, 0 for a leg with no stop set.
 *
 * This is what a "position count" cap misses: three separately-sized,
 * separately-stopped positions across correlated underlyings (NIFTY,
 * BANKNIFTY, FINNIFTY) can all breach in the same macro move, and the book's
 * real loss is the *sum* of their stops, not a single bounded trade.
 */
export function legRiskAtStop(leg: RiskLeg): number {
  if (leg.stop_loss === null) return 0;
  const perUnit =
    leg.side === "BUY" ? leg.avg_price - leg.stop_loss : leg.stop_loss - leg.avg_price;
  return Math.max(0, perUnit) * leg.quantity;
}

/** Total rupees at risk across every open leg, at each one's own stop. */
export function portfolioRiskAtStop(legs: RiskLeg[]): number {
  return legs.reduce((sum, leg) => sum + legRiskAtStop(leg), 0);
}
